src/main.cpp

```diff
@@ -6975,12 +6989,18 @@ bool static ProcessMessage(const CChainParams& chainparams, CNode* pfrom, string
 
     else if (strCommand == "verack")
     {
+        LOCK(cs_main);
+        CNodeState* state = State(pfrom->GetId());
+        if (state->fCurrentlyConnected) {
+            pfrom->PushMessage("reject", strCommand, REJECT_DUPLICATE, std::string("Duplicate verack message"));
+            Misbehaving(pfrom->GetId(), 1);
+            return false;
+        }
```

Comment:
I confirmed that `State()` is used elsewhere without being guarded by `pfrom->fNetworkNode` (which makes sense as it is intended for decoupling node-relevant network state from the network locks).

Currently the new reject will never trigger when a duplicate `verack` message is received from a non-`fNetworkNode`, which AFAICT includes all inbound connections (`fNetworkNode` is only set to true in `OpenNetworkConnection` for outbound connections). This implies that if we want to reject duplicate verack messages from all non-local `CNode`s (both outbound and inbound), then we need a new boolean to track that state (because `fCurrentlyConnected` is from upstream, and so we shouldn't go changing its semantics inside the node).

Reply:
I believe the presence, actually, of two separate booleans (`fCurrentlyConnected` from State and `fSuccessfullyConnected` from CNode member) are a bit confusing and, IMHO, misleading. Wrt specifically to the latter its meaning is ambiguous as it may appear as a redundant statement of what already is in State but at the same time it doesn't mean the handshake process has been fully completed (exchange amongst the two ends of the socket of **both** `version` and `verack`). 
I believe it would be beneficial to move the valuation of `fSuccessfullyConnected` to true only after `verack` message has been successfully parsed and processed **and** at the same time move there as well :

1. the `getaddr` request (if applicable)
2. the advertising of local address (if applicable)
3. the tagging as "good" (in addrman) of the sender's address (if applicable): actually a "good" address, IMHO, should be capable at least to complete the handshake

All of this should be helpful to solve issues #5168 and #6038 (protocol handshake should NOT be interleaved by any message other than version and verack).

Reply:
Also worth mentioning on the matter that latest bitcoin core implementation totally disregards the 26 bytes of the "from" address in version message while acknowledging the remote peer address from its addr messages.

Reply:
As #6038 demonstrates, we cannot drop messages that are received before we have sent "verack", because previous versions of zcashd send them and expect a response. We can stop sending them before the handshake has completed, but we must accept them when received.

Reply:
Anyway, #5168 and #6038 are a separate issue or issues that we should not attempt to fix in this PR, which is already complicated enough to analyse. I'm with @str4d that we should not be changing the semantics of variables that are present in Bitcoin Core.

Reply:
> As #6038 demonstrates, we cannot drop messages that are received before we have sent "verack", because previous versions of zcashd send them at that point and expect a response. We can stop sending them before we have received "verack", but we must accept them when received.

I think abstain from consuming them before the handshake has been completed in full could be a valuable intermediate solution. Anyhow, just for my knowledge and education, [latest bitcoin core's code base have moved the valuation of `fSuccessfullyConnected` flag from CNode's into the processing of `verack` message](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L3592-L3642).

In any case I'm not pushing to have more changes in this PR specifically.